ffmpeg/lpms_ffmpeg.c

```diff
@@ -1021,18 +1030,31 @@ int process_in(struct input_ctx *ictx, AVFrame *frame, AVPacket *pkt)
   // video frames, so continue on to audio.
 
   // Flush video decoder.
-  // To accommodate CUDA, we feed the decoder a a sentinel (flush) frame.
-  // Once the flush frame has been decoded, the decoder is fully flushed.
-  // TODO this is unnecessary for SW decoding! SW process should match audio
   if (ictx->vc) {
-    send_first_pkt(ictx);
-
-    ret = avcodec_receive_frame(ictx->vc, frame);
-    pkt->stream_index = ictx->vi;
-    if (!ret) {
-      if (is_flush_frame(frame)) ictx->flushed = 1;
-      return ret;
+    // Only use sentinel flushing for CUDA, and if we've started getting frames back
+    if (ictx->flush_mode == LPMS_FLUSH_UNDEFINED) {
+      ictx->flush_mode = (framecount > 0 && ictx->hw_type == AV_HWDEVICE_TYPE_CUDA) ? LPMS_FLUSH_SENTINEL : LPMS_FLUSH_DEFAULT;
+    }
+    if (ictx->flush_mode == LPMS_FLUSH_SENTINEL) {
+      // To accommodate CUDA, we feed the decoder a a sentinel (flush) frame.
+      // Once the flush frame has been decoded, the decoder is fully flushed.
+      send_first_pkt(ictx);
+
+      ret = avcodec_receive_frame(ictx->vc, frame);
+      if (!ret && is_flush_frame(frame))  ictx->flushed = 1;
+    } else {
+      // Classic flushing by sending NULL packets to decoder
+      ictx->flushing = 1;
+      ret = avcodec_send_packet(ictx->vc, NULL);
```

Comment:
Should `ret` be checked for errors here? If it doesn't need to be checked, maybe we can just not set it?

Reply:
ah yes it is not needed, i had it for debugging then forgot.